### ISSUE TITLE
module not found: community.ed#ed-logging_2.12;0.0.2

### ISSUE THREAD

**Issue by harinarayan2018** [debiki/talkyard#25]
Hi, 

While installing the dev version of the app getting this issue "module not found: community.ed#ed-logging_2.12;0.0.2" and this makes my play framework down and unable to run the app. 

Thank for your help in advance. Thanks

**Comment by kajmagnus** [comment]
Look at step 3 in [the getting started instructions](https://github.com/debiki/talkyard#the-instructions):

> 3\. Compile and SBT publish-local a logging library.
> 
>  sudo s/d-cli
>  project ed-logging
>  publish-local
>  \# then CTRL+D to exit

Unfortunately, seems it's impossible to have Play compile this logging library, in development mode, before Play tries to load it already. Play's internal logging things, start logging messages, before Play would start compiling this lib.

So therefore it has to be done in a separate step, before starting Play.

**Comment by harinarayan2018** [comment]
Thanks, 
But I have already done with the step 3.  While doing step 3 there no error take place. After it when I do CTRL+D and compose docker then facing this issue. 
app_1      | [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
app_1      | [warn] 	::          UNRESOLVED DEPENDENCIES         ::
app_1      | [warn] 	::::::::::::::::::::::::::::::::::::::::::::::
app_1      | [warn] 	:: community.ed#ed-logging_2.12;0.0.2: not found
app_1      | [warn] 	::::::::::::::::::::::::::::::::::::::::::::::

and container of app go down. :(

**Comment by kajmagnus** [comment]
Ok I'll try to install the development repo on a new machine, maybe this afternoon but more likely tomorrow.

Operating system and version?

Can you run these commands & copy-paste the output? (so I'll see e.g. which git revision, & software versions)

```
docker -v
docker-compose -v
uname -a  # optionally, edit out your laptop's network name
cat /etc/issue
git log --oneline -n10
git status
```

And, unless you've added something private somewhere:

```
git diff
```
(maybe the diff will be empty? Or, if it's too long, maybe paste in a Gist instead?)

**Comment by harinarayan2018** [comment]
ubuntu 16.04
Docker version 18.03.1-ce, build 9ee9f40
docker-compose version 1.18.0, build 8dd22a9
Linux talkyard-new 4.4.0-122-generic #146-Ubuntu SMP Mon Apr 23 15:34:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by kajmagnus** [comment]
Maybe the problem is that nowadays a newer SBT version is in use, requiring you to type the commands slightly differently. There ought to have been error messages about this. And I should update the docs.

Instead of `project ed-logging` and `publish-local`,  
one needs to type `project edLogging` and `publishLocal`.

I test installed in Ubuntu 16.04 in a Vagrant VM. Worked fine for me, with the two changes mentioned above.

Can you copy-paste the output from this?: (so we'll see if there're any files published or not, in your case)

```
vagrant@ubuntu-xenial:~$ find ~/.ivy2/ -iname '*ed-logging*' 
```

When I run that `find`, I see: (and this shows that ed-logging has been built & published locally in my Vagrant vm)

```
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/srcs/ed-logging_2.12-sources.jar.md5
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/srcs/ed-logging_2.12-sources.jar.sha1
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/srcs/ed-logging_2.12-sources.jar
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/docs/ed-logging_2.12-javadoc.jar
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/docs/ed-logging_2.12-javadoc.jar.sha1
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/docs/ed-logging_2.12-javadoc.jar.md5
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/poms/ed-logging_2.12.pom
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/poms/ed-logging_2.12.pom.md5
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/poms/ed-logging_2.12.pom.sha1
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/jars/ed-logging_2.12.jar.md5
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/jars/ed-logging_2.12.jar
/home/vagrant/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/jars/ed-logging_2.12.jar.sha1
```

Can you try building ed-logging again? Like so:  (note: type  `project edLogging` and `publishLocal` instead of `project ed-logging` and `publish-local` — I'll update the docs to reflect this)

```
/talkyard$ sudo s/d-killdown   # stops Talkyard if running already
/talkyard$ sudo s/d-cli
(... Docker log messages ...)
[talkyard-server] $
[talkyard-server] $ project ed-logging     <— ooops won't work
[error] Not a valid project ID: ed-logging (similar: edLogging)
[error] project ed-logging
[error]                   ^
[talkyard-server] $ 
[talkyard-server] $ project edLogging     <—— this works, because new version of SBT
[info] Set current project to ed-logging (in build file:/opt/talkyard/app/)
sbt:ed-logging> 
sbt:ed-logging> 
sbt:ed-logging> publish-local    <— oops won't work
[error] Not a valid command: publish-local
[error] Not a valid project ID: publish-local
[error] Expected ':'
[error] Not a valid key: publish-local (similar: publishLocal, publishTo, publish)
[error] publish-local
[error]              ^
sbt:ed-logging> 
sbt:ed-logging> publishLocal     <—— this works
[info] Packaging /opt/talkyard/app/modules/ed-logging/target/scala-2.12/ed-logging_2.12-0.0.2-sources.jar ...
[info] Updating ...
[info] Done packaging.
... downloading lots of stuff ...
[info] Done updating.
[info] Wrote /opt/talkyard/app/modules/ed-logging/target/scala-2.12/ed-logging_2.12-0.0.2.pom
[info] Main Scala API documentation to /opt/talkyard/app/modules/ed-logging/target/scala-2.12/api...
[info] Compiling 1 Scala source to /opt/talkyard/app/modules/ed-logging/target/scala-2.12/classes ...
[info] Non-compiled module 'compiler-bridge_2.12' for Scala 2.12.3. Compiling...
[info]   Compilation completed in 12.909s.
model contains 4 documentable templates
[info] Done compiling.
[info] Main Scala API documentation successful.
[info] Packaging /opt/talkyard/app/modules/ed-logging/target/scala-2.12/ed-logging_2.12-0.0.2-javadoc.jar ...
[info] Packaging /opt/talkyard/app/modules/ed-logging/target/scala-2.12/ed-logging_2.12-0.0.2.jar ...
[info] Done packaging.
[info] Done packaging.
[info] :: delivering :: community.ed#ed-logging_2.12;0.0.2 :: 0.0.2 :: release :: Thu May 10 13:14:49 GMT 2018
[info] 	delivering ivy file to /opt/talkyard/app/modules/ed-logging/target/scala-2.12/ivy-0.0.2.xml
[info] 	published ed-logging_2.12 to /home/owner/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/poms/ed-logging_2.12.pom
[info] 	published ed-logging_2.12 to /home/owner/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/jars/ed-logging_2.12.jar
[info] 	published ed-logging_2.12 to /home/owner/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/srcs/ed-logging_2.12-sources.jar
[info] 	published ed-logging_2.12 to /home/owner/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/docs/ed-logging_2.12-javadoc.jar
[info] 	published ivy to /home/owner/.ivy2/local/community.ed/ed-logging_2.12/0.0.2/ivys/ivy.xml
[success] Total time: 27 s, completed May 10, 2018 1:14:50 PM
```

**Comment by harinarayan2018** [comment]
Hi, 

Thanks for your help, As you mentioned I already made that changes as when we run "project ed-logging" it gives suggestion for project edLogging. 

When I run - find ~/.ivy2/ -iname '*ed-logging*' I found nothing. Well I'm planning to do all steps again with fresh setup. 

I will let you know if it works for me. Thanks

**Comment by harinarayan2018** [comment]
Below is the output of the command, I have created new remote server of digital ocean with 4GB RAM and try out the all the steps. All steps run successful but still the same error.

```root@talkyard-new-v1:/opt/talkyard# sudo docker-compose run --rm app   find /home/owner/ -iname '*ed-logging*'
Starting edd_rdb_1 ... done
Starting edd_cache_1 ... done
Starting edd_search_1 ... done
Running the Play CMD:
+ exec su -c 'find /home/owner/ -iname *ed-logging*' root
```


```root@talkyard-new-v1:/opt/talkyard# sudo docker-compose run --rm app   ls -al /home/owner/.ivy2/  /home/owner/.sbt/
Starting edd_rdb_1 ... done
Starting edd_search_1 ... 
Starting edd_cache_1 ... done
Running the Play CMD:
+ exec su -c 'ls -al /home/owner/.ivy2/ /home/owner/.sbt/' root
/home/owner/.ivy2/:
total 8
drwxr-xr-x    2 root     root          4096 May 11 06:31 .
drwxr-xr-x    4 root     root          4096 May 11 08:54 ..

/home/owner/.sbt/:
total 8
drwxr-xr-x    2 root     root          4096 May 11 06:31 .
drwxr-xr-x    4 root     root          4096 May 11 08:54 ..
```

**Comment by kajmagnus** [comment]
Hmm, you're `root` on the Digital Ocean host? I remember running into problems, when I tried to run the development environment, in a VM, as root, but didn't investigate much further, because I expected people to install the dev stuff, on their laptops as not-root...

... So would be interesting to understand your use case :- )  
**What's the reasons you're installing the dev env, in a remote VM?** (and thereby happen to become root)  
E.g. to little memory, on the laptop? ... Or ... you really have in mind to install Talkyard's *development* Git repo? and not do a [*production* installation](https://github.com/debiki/talkyard-prod-one)? Or maybe you use a VM as a safety measure, in case of bugs or bad code, in the dev repo, that might mess up your laptop?

I can look into making running the dev stuff as `root` work. Being able to use a DO VM for development seems like a good thing to me.

Here's why I think you're root, in the DO VM. From the app container's Docker [entrypoint.sh](https://github.com/debiki/talkyard/blob/master/docker/app-dev/entrypoint.sh) file:

```
# Create user 'owner' with the same id as the person who runs docker, so that [...]
file_owner_id=`ls -adn | awk '{ print $3 }'`

...

echo Running the Play CMD:

if [ $file_owner_id -ne 0 ] ; then    <—  -ne 0 means not-equal-to-0, and 0 = root
  # Prevent a file-not-found exception in case ~/.ivy2 and ~/.sbt didn't exist, so ...
  ...
  set -x
  exec su -c "$*" owner     <—— look: "exec ..."  ends with  "owner"
else
  # We're root (user id 0), both on the Docker host and here in the container.
  # `exec su ...` is the only way I've found that makes Yarn and Gulp respond to CTRL-C,
  # so using `su` here although we're root already.
  set -x
  exec su -c "$*" root     <—— only ends with "root", if you're root on the host
fi
```

When you're root, I think SBT will store the libs it has downloaded, into `/root/.ivy2` and `/root/.sbt` instead of in `/home/owner/` (which mounted on the host, and persists across container deletion & recreation). All libs below `/root/` disappear, whenever the app container gets recreated. Right now I think that's what has happened in your case.

**Comment by harinarayan2018** [comment]
Hi, 
The reason to being the root is - I and my friend want to make some changes as per our need and both are on different location. Will also notify you about the changes we are going to make so if you like you can persist. 

I tried to create a user on the same VM and tried to install the application by that user but no success yet. I am not getting why it is happening but you are the right one who can suggest what to do next. 

Do I need to make some changes to so that we can install dev on VM ? 

We are thankful for your suggestion and help.

**Comment by kajmagnus** [comment]
Hello! ok, interesting use case. You both share the same root account, for simplicity, if I understood correctly.

I think I've made the dev env work for `root` now (should only be used inside a dedicated vm obviously :- )). If you do this:

```
git fetch origin
git checkout master
git status
```

and if there're no important changes to save, then:
```
git reset --hard origin/master
```

Then start at step 3 again, i.e. *"Compile and SBT publishLocal ... "*.

What I did, was to set `HOME=/home/owner` and, for Java, `-Duser.home=/home/owner`. This makes the same `/home/owner/.sbt` and `.ivy` cache dirs get used, regardless of if the Docker host user is root or another "normal" user.

Plz tell me if it works for you?
